File: flutter_app/lib/location_store.dart
import 'package:flutter/foundation.dart';
import 'package:geolocator/geolocator.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'locale_store.dart';
import 'viloyat.dart';

enum LocationStatus { idle, loading, granted, denied, unavailable }

/// GPS orqali foydalanuvchi qaysi viloyatda ekanini aniqlaydi — o'sha
/// viloyatdagi (yoki viloyati ko'rsatilmagan) firmalarning mahsulotlarini
/// ko'rsatish uchun (`Product.company_viloyat` filtri, backend `?viloyat=`).
/// Foydalanuvchi xohlasa katalogdan qo'lda "Barchasi"ga o'tishi ham mumkin.
class LocationStore extends ChangeNotifier {
  static const _prefKey = 'selected_viloyat';
  static const _latKey = 'gps_lat';
  static const _lngKey = 'gps_lng';

  String? _viloyat; // null = "Barchasi" (filtrsiz)
  double? _lat;
  double? _lng;
  LocationStatus _status = LocationStatus.idle;

  String? get viloyat => _viloyat;
  // Aniq GPS koordinatasi — mavjud bo'lsa, firma xizmat radiusi bo'yicha
  // filtrlashda viloyat o'rniga shu ishlatiladi (qarang apps/products/views.py
  // `lat`/`lng` parametri, backend/common/geo.py).
  double? get lat => _lat;
  double? get lng => _lng;
  LocationStatus get status => _status;
  String viloyatLabelText(LocaleStore loc) => viloyatLabel(_viloyat, loc);

  Future<void> init() async {
    final prefs = await SharedPreferences.getInstance();
    final saved = prefs.getString(_prefKey);
    final savedLat = prefs.getDouble(_latKey);
    final savedLng = prefs.getDouble(_lngKey);
    if (saved != null) {
      _viloyat = saved;
      _lat = savedLat;
      _lng = savedLng;
      _status = LocationStatus.granted;
      notifyListeners();
      return;
    }
    await detectFromGps();
  }

  Future<void> detectFromGps() async {
    _status = LocationStatus.loading;
    notifyListeners();
    try {
      if (!await Geolocator.isLocationServiceEnabled()) {
        _status = LocationStatus.unavailable;
        notifyListeners();
        return;
      }
      var permission = await Geolocator.checkPermission();
      if (permission == LocationPermission.denied) {
        permission = await Geolocator.requestPermission();
      }
      if (permission == LocationPermission.denied ||
          permission == LocationPermission.deniedForever) {
        _status = LocationStatus.denied;
        notifyListeners();
        return;
      }
      final position = await Geolocator.getCurrentPosition(
        locationSettings: const LocationSettings(
          accuracy: LocationAccuracy.low,
        ),
      );
      final nearest = nearestViloyat(position.latitude, position.longitude);
      await setViloyat(nearest.code, lat: position.latitude, lng: position.longitude);
      _status = LocationStatus.granted;
    } catch (_) {
      _status = LocationStatus.unavailable;
    }
    notifyListeners();
  }

  Future<void> setViloyat(String? code, {double? lat, double? lng}) async {
    _viloyat = code;
    // Qo'lda boshqa viloyat tanlansa, avvalgi GPS koordinatasi endi noto'g'ri
    // bo'lib qoladi — shuning uchun faqat GPS orqali kelgan chaqiruvdagina
    // (lat/lng berilganda) saqlanadi, aks holda tozalanadi.
    _lat = lat;
    _lng = lng;
    notifyListeners();
    final prefs = await SharedPreferences.getInstance();
    if (code == null) {
      await prefs.remove(_prefKey);
    } else {
      await prefs.setString(_prefKey, code);
    }
    if (lat != null && lng != null) {
      await prefs.setDouble(_latKey, lat);
      await prefs.setDouble(_lngKey, lng);
    } else {
      await prefs.remove(_latKey);
      await prefs.remove(_lngKey);
    }
  }
}

File: ios/FurniturePlatform/Sources/Networking/LocationStore.swift
import CoreLocation
import Foundation

enum LocationStatus {
    case idle, loading, granted, denied, unavailable
}

/// GPS orqali foydalanuvchi qaysi viloyatda ekanini aniqlaydi — o'sha
/// viloyatdagi (yoki viloyati ko'rsatilmagan) firmalarning mahsulotlarini
/// ko'rsatish uchun (`Product.companyViloyat` filtri, backend `?viloyat=`).
/// Flutter'dagi `lib/location_store.dart` bilan bir xil naqsh.
@MainActor
final class LocationStore: NSObject, ObservableObject, CLLocationManagerDelegate {
    private static let prefKey = "selected_viloyat"
    private static let latKey = "gps_lat"
    private static let lngKey = "gps_lng"

    @Published private(set) var viloyat: String?
    // Aniq GPS koordinatasi — mavjud bo'lsa, firma xizmat radiusi bo'yicha
    // filtrlashda viloyat o'rniga shu ishlatiladi (qarang backend
    // apps/products/views.py `lat`/`lng` parametri, backend/common/geo.py).
    @Published private(set) var lat: Double?
    @Published private(set) var lng: Double?
    @Published private(set) var status: LocationStatus = .idle

    private let manager = CLLocationManager()

    var viloyatLabelText: String { viloyatLabel(viloyat) }

    override init() {
        super.init()
        manager.delegate = self
    }

    func bootstrap() {
        if let saved = UserDefaults.standard.string(forKey: Self.prefKey) {
            viloyat = saved
            if UserDefaults.standard.object(forKey: Self.latKey) != nil {
                lat = UserDefaults.standard.double(forKey: Self.latKey)
                lng = UserDefaults.standard.double(forKey: Self.lngKey)
            }
            status = .granted
            return
        }
        detectFromGps()
    }

    func detectFromGps() {
        status = .loading
        let authStatus = manager.authorizationStatus
        switch authStatus {
        case .notDetermined:
            manager.requestWhenInUseAuthorization()
        case .denied, .restricted:
            status = .denied
        case .authorizedWhenInUse, .authorizedAlways:
            manager.requestLocation()
        @unknown default:
            status = .unavailable
        }
    }

    func setViloyat(_ code: String?, lat: Double? = nil, lng: Double? = nil) {
        viloyat = code
        // Qo'lda boshqa viloyat tanlansa, avvalgi GPS koordinatasi endi
        // noto'g'ri bo'lib qoladi — shuning uchun faqat GPS orqali kelgan
        // chaqiruvdagina (lat/lng berilganda) saqlanadi, aks holda tozalanadi.
        self.lat = lat
        self.lng = lng
        if let code {
            UserDefaults.standard.set(code, forKey: Self.prefKey)
        } else {
            UserDefaults.standard.removeObject(forKey: Self.prefKey)
        }
        if let lat, let lng {
            UserDefaults.standard.set(lat, forKey: Self.latKey)
            UserDefaults.standard.set(lng, forKey: Self.lngKey)
        } else {
            UserDefaults.standard.removeObject(forKey: Self.latKey)
            UserDefaults.standard.removeObject(forKey: Self.lngKey)
        }
    }

    nonisolated func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        Task { @MainActor in
            switch manager.authorizationStatus {
            case .authorizedWhenInUse, .authorizedAlways:
                manager.requestLocation()
            case .denied, .restricted:
                status = .denied
            default:
                break
            }
        }
    }

    nonisolated func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        guard let loc = locations.last else { return }
        let nearest = nearestViloyat(lat: loc.coordinate.latitude, lng: loc.coordinate.longitude)
        Task { @MainActor in
            setViloyat(nearest.code, lat: loc.coordinate.latitude, lng: loc.coordinate.longitude)
            status = .granted
        }
    }

    nonisolated func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        Task { @MainActor in status = .unavailable }
    }
}
